Repro

abhishek-das-gupta opened the pull request.

anacrolix commented:
Thanks for the input. If I'm reading your log however, it works fine for you? I need it to demonstrate failure so I can find what is wrong.

abhishek-das-gupta commented:
> Thanks for the input. If I'm reading your log however, it works fine for you? I need it to demonstrate failure so I can find what is wrong.

@anacrolix  Thank you very much for providing this test case! But I'm not sure what do you mean by "failure".  There is no failure per se. The #953 issue is related to slow performance not failure. 

If you look at `tests/peers-bootstrapping/Client-Torrent.log` in the PR, you can see only Client0 making progress while other Client[1..5] are all stuck at 0% progress until the Client0 reaches 100%. This is what I described in #953. I added the full-status of all the node in the [comment](https://github.com/anacrolix/torrent/discussions/953#discussioncomment-9889190)

Although I'm not sure if the [changes](https://github.com/anacrolix/torrent/pull/956/commits/be1ca18e086879a53405504ba46b45f7d64de9cf#diff-4c771f4709ac57dee519b205f07bfcee5a4f0d013cb3dd0cda509eddfd2fb9ccR107-R108) that I added in `main.go` in commit1 are correct or not. I guess since I made the `initalTorrent` completion as a blocking call due to which other peers are not progressing?


I've added another [commit](https://github.com/anacrolix/torrent/pull/956/commits/ddac974aa6f348e2ae30b72583285d422f03d10f) right now, which seems to be as closes to do what we are doing in our code.



```
source dir <---- is where the node0 downloads "file" binary.
client0 dir<----- is the cfg.DefaultStorage directory for the node0 or client0
client1    ----
.                    |
.                    | -> all the clients add the WebSeed called file://source/file (In my code, it would be a HTTP url )
.                    |
client5 -----
```

When I make these changes, the test main.go seems to stuck. I'm unable to debug this? Can you please provide insights what am I doing wrong? Thanks

anacrolix commented:
Thanks! I'll take a look

anacrolix commented:
Well I spotted a masked bug in the webseed code. However that's not the issue.

The things I fixed:

You can't use file:// for a webseed URL. Spin up a HTTP server if you want to do that (do it in Go, in the repro, easy).

You can't not seed the actual completed data if webseeding isn't available.

You can't copy over the storage files while they're in use. You need to do it before the torrent's storage is active.

At least one client needs to have the actual full metainfo. You can't get this over webseeding, you need to use the metainfo.

abhishek-das-gupta commented:
Thanks @anacrolix  for taking a look and providing the pointers.

I mimicked your testcase into the codebase and the operations into the following sequence:

#### Activate the torrent on each node/client/peer:
The following code is the client written in python that runs on each agent node which calls the  `activate-torrent` API served by out `TORRENT_SERVER`
```python
LOG.info("Activate Torrent...%s", torrent_name)
response = requests.get(f"{self.TORRENT_SERVER}/activate-torrent?fileName={torrent_name}")
response.raise_for_status()
LOG.info("Activated Torrent...%s", torrent_name)
```

The above API calls the `TORRENT_SERVER` which is written in GO that communicates with anacrolix/torrent lib
```go
func (cl *Client) ActivateTorrent(torrentName string) error {
	filePath := filepath.Join(cl.storagePath, torrentName)
	if !strings.HasSuffix(filePath, TORRENT_FILE) {
		filePath += TORRENT_FILE
	}
	mi, err := metainfo.LoadFromFile(filePath)
	if err != nil {
		log.Logger.Error("Couldn't load meta-info file", err)
		return err
	}
	t, err := cl.AddTorrent(mi)
	if err != nil {
		log.Logger.Info(err)
		return err
	}
	<-t.GotInfo()
	log.Logger.Infof("Activation of torrent %s was successful with t.info.BestName(): %s", torrentName, t.Info().BestName())
	cl.torrents[torrentName] = TorrentInfo{
		hash:    mi.HashInfoBytes(),
		created: time.Now(),
		expiry:  time.Now().Add(cl.torrentExpiry),
	}
	return nil
}
```

#### Next step in the sequence is : Adding peers for each node:
Again, the client python code which makes an API call called `add-peers` to `TORRENT_SERVER` 
```python
        LOG.info("Adding rack peers %s for torrent %s...", self.rackpeers, torrent_name)
        data = {"torrent" : torrent_name, "peers" : list(self.rackpeers)}
        res = requests.post(f"{self.FLOODIX_SERVER}/add-peers",json=data)
        res.raise_for_status()
        LOG.info("Added rack peers %s for torrent %s...", self.rackpeers, torrent_name)
        is_torrent_active = True
```


The Go code of the `TORRENT_SERVER` that communicates with anacrolix/torrent:
```go
func (cl *Client) AddPeers(torrentName string, peers []string) error {
	t, err := cl.getTorrent(torrentName)
	if err != nil {
		return err
	}
	n := t.AddPeers(createPeers(peers, cl.configs.Port))
	log.Logger.Infof("Added %d new peers", n)
	return nil

}

func createPeers(addrs []string, port int) (ret []torrent.PeerInfo) {
	for _, ta := range addrs {
		addr := ta
		if !strings.ContainsRune(ta, ':') {
			addr = fmt.Sprintf("%s:%d", ta, port)
		}
		ret = append(ret, torrent.PeerInfo{
			Addr:    stringAddr(addr),
			Source:  torrent.PeerSourceDirect,
			Trusted: true,
		})
	}
	return ret
}
```

#### Third step, is starting the download on each node after the first two steps:
```python
LOG.info("Distributing torrent %s...", torrent_name)
response = requests.get(f"{self.FLOODIX_SERVER}/distribute?fileName={torrent_name}")
response.raise_for_status()
```

```go
func (cl *Client) DownloadFromTorrent(torrentName string, url string) error {
	_, ok := cl.torrents[torrentName]
	if ok {
		err := cl.AddAWebSeed(torrentName, url)
		if err != nil {
			return err
		}
		return nil
	}
	t, err := cl.getTorrent(torrentName)
	if err != nil {
		return err
	}
	t.DownloadAll()
	return nil
}
```

Lastly I just loop to poll the status from our python code:
```python
      while True:
        # Query the status from your Golang server
        response = requests.get(f"{self.TORREN_SERVER}/status?fileName={torrent_name}")
        response.raise_for_status()
```


I seeing the master node's download completes quickly but other nodes are stuck at 0 `Pieces Completed`
I checked the status of each node 

I checked the node#1 status and it doesn't have any peers even though `add-peers` API was called and it was successful.
```
(torrent.TorrentStats) {
 ConnStats: (torrent.ConnStats) {
  BytesWritten: (torrent.Count) 737259,
  BytesWrittenData: (torrent.Count) 0,
  BytesRead: (torrent.Count) 4788,
  BytesReadData: (torrent.Count) 0,
  BytesReadUsefulData: (torrent.Count) 0,
  BytesReadUsefulIntendedData: (torrent.Count) 0,
  ChunksWritten: (torrent.Count) 0,
  ChunksRead: (torrent.Count) 0,
  ChunksReadUseful: (torrent.Count) 0,
  ChunksReadWasted: (torrent.Count) 0,
  MetadataChunksRead: (torrent.Count) 0,
  PiecesDirtiedGood: (torrent.Count) 0,
  PiecesDirtiedBad: (torrent.Count) 0
 },
 TotalPeers: (int) 0,     <---------------
 PendingPeers: (int) 0,
 ActivePeers: (int) 0, <------------------
 ConnectedSeeders: (int) 0,
 HalfOpenPeers: (int) 0,
 PiecesComplete: (int) 26950
}
webseeds:
0 peer conns:
```


While for Other nodes 2...5
```
 TotalPeers: (int) 5, <------ 
 PendingPeers: (int) 9, <-----
 ActivePeers: (int) 0,
 ConnectedSeeders: (int) 0,
 HalfOpenPeers: (int) 0,
 PiecesComplete: (int) 0
}
webseeds:
- https://ccycloud-1.t-bep.root.comops.site:7183/cmf/parcel/download/CDH-7.2.18-1.cdh7.2.18.p300.55014068-el8.parcel
  last unhandled error: never
  bep40-prio: fa2f2406
  last msg: never, connected: never, last helpful: never, itime: 0s, etime: 0s
  26950/26950 completed, 0 pieces touched, good chunks: 0/0:0 reqq: 0+0/(1/128):0/1024, flags: :WS:, dr: 0.0 KiB/s
  requested pieces:
0 peer conns: <-----------
```


Why the  peer#1 doesn't have any peer connection meanwhile other peer show they are waiting on the connection? What did I do wrong here? Can you please take a look

anacrolix commented:
Could you make the Go code reproduce this? I can't see anything wrong with your code at a glance, but it's not straightforward.

abhishek-das-gupta commented:
Unfortunately I'm unable to reproduce this in a test case.

However I've added some custom logs in `torrent.addPeers()` function. I've highlighted them using the arrow `<---` in the below code snippet:
```go
func (t *Torrent) addPeer(p PeerInfo) (added bool) {
	cl := t.cl
	torrent.Add(fmt.Sprintf("peers added by source %q", p.Source), 1)
	if t.closed.IsSet() {
		t.logger.Levelf(log.Debug, "Torrent is closed. Hence %v not added", p.addr()) <----
		return false
	}
	if ipAddr, ok := tryIpPortFromNetAddr(p.Addr); ok {
		if cl.badPeerIPPort(ipAddr.IP, ipAddr.Port) {
			torrent.Add("peers not added because of bad addr", 1)
			t.logger.Levelf(log.Debug, "Bad Ip %v and  port %v. Hence %v not added", ipAddr.IP, ipAddr.Port, p.addr()) <--
			// cl.logger.Printf("peers not added because of bad addr: %v", p)
			return false
		}
	}
	if replaced, ok := t.peers.AddReturningReplacedPeer(p); ok {
		log.Levelf(log.Debug, "Peer retuning replaced %v", replaced)<---
		torrent.Add("peers replaced", 1)
		if !replaced.equal(p) {
			t.logger.WithDefaultLevel(log.Debug).Printf("added %v replacing %v", p, replaced)
			added = true
		}
	} else {
		added = true
	}
	t.openNewConns()
	for t.peers.Len() > cl.config.TorrentPeersHighWater {
		t.logger.Levelf(log.Debug, "t.peer.Len(): %d is greater than cl.config.TorrentPeersHighWater:%d",
			t.peers.Len(), cl.config.TorrentPeersHighWater) <----
		_, ok := t.peers.DeleteMin()
		if ok {
			torrent.Add("excess reserve peers discarded", 1)
		}
	}
	t.logger.Levelf(log.Debug, "Final value of added: %v for peer %v", added, p.addr) <---
	return
}
```

The above  function is called through the following wrapper code:
```go
func (cl *Client) AddPeers(torrentName string, peers []string) error {
	t, err := cl.getTorrent(torrentName)
	if err != nil {
		return err
	}
	log.Logger.Infof("Check whether torrent is closed t.Closed(): %v", t.Closed())
	n := t.AddPeers(createPeers(peers, cl.configs.Port)) <---
	log.Logger.Infof("Added %d new peers", n) <---
	return nil

}
```


After addding the custom logs in `torrent.addPeers()` I see that node#1 (master node ) is not adding any peers
```
2024-07-05T14:47:40.036Z	info	log/logger.go:101	["GET http://localhost:7190/activate-torrent?fileName=CDH-7.2.18-1.cdh7.2.18.p300.55014068-el8.parcel HTTP/1.1" from 127.0.0.1:60524 - 200 70B in 289.861206ms]
2024-07-05T14:47:40.043Z	info	http/server.go:332	{CDH-7.2.18-1.cdh7.2.18.p300.55014068-el8.parcel []}
2024-07-05T14:47:40.043Z	info	floodix/server.go:319	Check whether torrent is closed t.Closed(): 0xc000294240
2024-07-05T14:47:40.043Z	info	floodix/server.go:321	Added 0 new peers
```

It is logging `Added 0 new peers` peers and not logging any of the custom logs that I added in the `torrent.addPeers()`


whereas other nodes{2...5} log the following:
Here is a snippet from node2
```
2024-07-05T14:47:54.174Z	info	log/logger.go:101	["GET http://localhost:7190/status?fileName=CDH-7.2.18-1.cdh7.2.18.p300.55014068-el8.parcel HTTP/1.1" from 127.0.0.1:32804 - 200 250B in 2.96738ms]
2024-07-05T14:47:54.826Z	info	http/server.go:332	{CDH-7.2.18-1.cdh7.2.18.p300.55014068-el8.parcel [10.140.86.9:7191 10.140.134.71:7191 10.140.230.7:7191 10.140.85.142:7191 10.140.177.68:7191]}
2024-07-05T14:47:54.826Z	info	floodix/server.go:319	Check whether torrent is closed t.Closed(): 0xc00030e180
2024-07-05T14:47:54.826Z	debug	log/logger.go:157	Final value of added: true for peer 0xa7e020  [ github.com/anacrolix/torrent   torrent.go:363 ] 

2024-07-05T14:47:54.826Z	debug	log/logger.go:157	Final value of added: true for peer 0xa7e020  [ github.com/anacrolix/torrent   torrent.go:363 ] 

2024-07-05T14:47:54.826Z	debug	log/logger.go:157	Final value of added: true for peer 0xa7e020  [ github.com/anacrolix/torrent   torrent.go:363 ] 

2024-07-05T14:47:54.826Z	debug	log/logger.go:157	Final value of added: true for peer 0xa7e020  [ github.com/anacrolix/torrent   torrent.go:363 ] 

2024-07-05T14:47:54.826Z	debug	log/logger.go:157	Final value of added: true for peer 0xa7e020  [ github.com/anacrolix/torrent   torrent.go:363 ] 

2024-07-05T14:47:54.826Z	info	floodix/server.go:321	Added 5 new peers
```

Here is a snippet of logs from nod#4
```
2024-07-05T14:47:54.764Z	info	http/server.go:332	{CDH-7.2.18-1.cdh7.2.18.p300.55014068-el8.parcel [10.140.86.9:7191 10.140.134.71:7191 10.140.230.7:7191 10.140.85.142:7191 10.140.177.68:7191]}
2024-07-05T14:47:54.764Z	info	floodix/server.go:319	Check whether torrent is closed t.Closed(): 0xc000294060
2024-07-05T14:47:54.764Z	debug	log/logger.go:157	Final value of added: true for peer 0xa7e020  [ github.com/anacrolix/torrent   torrent.go:363 ] 

2024-07-05T14:47:54.764Z	debug	log/logger.go:157	Final value of added: true for peer 0xa7e020  [ github.com/anacrolix/torrent   torrent.go:363 ] 

2024-07-05T14:47:54.764Z	debug	log/logger.go:157	Final value of added: true for peer 0xa7e020  [ github.com/anacrolix/torrent   torrent.go:363 ] 

2024-07-05T14:47:54.764Z	debug	log/logger.go:157	Final value of added: true for peer 0xa7e020  [ github.com/anacrolix/torrent   torrent.go:363 ] 

2024-07-05T14:47:54.764Z	debug	log/logger.go:157	Final value of added: true for peer 0xa7e020  [ github.com/anacrolix/torrent   torrent.go:363 ] 

2024-07-05T14:47:54.764Z	info	floodix/server.go:321	Added 5 new peers
```


Any pointers why the node#1 is not adding the rest of the peers and also not logging any of the custom logs? It alteast should have logged
```go
t.logger.Levelf(log.Debug, "Final value of added: %v for peer %v", added, p.addr) <---
```

abhishek-das-gupta commented:
@anacrolix unfortunately I'm unable to repro it in the test case.

I have added more [logs](https://github.com/abhishek-das-gupta/torrent/commit/c2335203f439fa2f15b4326b8387d777d1c9ce32) in `t.addPeers()` and `t.openNewConns()`  and I've captured the logs for node1 and node2 where node1 completes quickly but there isn't any connection happening b/w node1 or node{2...5}

In the screenshot you can see as well where that only node1 has distributed while other nodes have no progress. 
<img width="1162" alt="Screenshot 2024-07-09 at 1 11 35 PM" src="https://github.com/anacrolix/torrent/assets/33230771/eb1e84c9-9f10-4b74-9cff-fb52f2207ec6">


For node2, I can see the following custom log that I've added
```

2024-07-09T00:22:54.305-0700	debug	log/logger.go:157	#Floodix: Opening connection with peer 10.65.169.72:7191...  [ github.com/anacrolix/torrent   torrent.go:354 ] 

2024-07-09T00:22:54.305-0700	debug	log/logger.go:157	#Floodix: Inside openNewConns() function  [ github.com/anacrolix/torrent   torrent.go:1576 ] 

2024-07-09T00:22:54.305-0700	debug	log/logger.go:157	#Floodix: t.peers.Len(): 7  [ github.com/anacrolix/torrent   torrent.go:1578 ] 

2024-07-09T00:22:54.305-0700	debug	log/logger.go:157	#Floodix: torrent doesn't want outgoing connections since !t.wantOutgoingConns(): true  [ github.com/anacrolix/torrent   torrent.go:1581 ] 

2024-07-09T00:22:54.306-0700	debug	log/logger.go:157	#Floodix: Final value of added: false for peer 0xa80660  [ github.com/anacrolix/torrent   torrent.go:364 ] 
```

The log I see getting spammed in node2's log file:
```
2024-07-09T00:22:54.305-0700	debug	log/logger.go:157	#Floodix: torrent doesn't want outgoing connections since !t.wantOutgoingConns(): true  [ github.com/anacrolix/torrent   torrent.go:1581 ] 
```

while the nod#1 is trying to make connections?
```
2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Inside t.AddPeers() the peerInfos are [{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.169.72:7191 M false 0 true} {[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.170.47:7191 M false 0 true} {[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.140.160:7191 M false 0 true} {[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.138.33:7191 M false 0 true} {[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.131.12:7191 M false 0 true}]  [ github.com/anacrolix/torrent   t.go:245 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Opening connection with peer 10.65.169.72:7191...  [ github.com/anacrolix/torrent   torrent.go:354 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Inside openNewConns() function  [ github.com/anacrolix/torrent   torrent.go:1576 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: t.peers.Len(): 1  [ github.com/anacrolix/torrent   torrent.go:1578 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: t.peers.PopMax(): 10.65.169.72:7191  [ github.com/anacrolix/torrent   torrent.go:1597 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: initiating connection with peer 10.65.169.72:7191 with opts: {{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.169.72:7191 M false 0 true} 0xc000208a88 false false false {false true}}  [ github.com/anacrolix/torrent   torrent.go:1606 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Inside initateConn() with opts: {{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.169.72:7191 M false 0 true} 0xc000208a88 false false false {false true}}  [ github.com/anacrolix/torrent   torrent.go:2730 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Not ignoring Limits since !ignoreLimits true  [ github.com/anacrolix/torrent   torrent.go:2743 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: calling go t.cl.outgoingConnection({{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.169.72:7191 M false 0 true} 0xc000208a88 false false false {false true}}, &{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.169.72:7191 M false 0 true})  [ github.com/anacrolix/torrent   torrent.go:2756 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Connection initiated with peer 10.65.169.72:7191 with opts: {{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.169.72:7191 M false 0 true} 0xc000208a88 false false false {false true}}  [ github.com/anacrolix/torrent   torrent.go:1609 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Final value of added: true for peer 0xa80660  [ github.com/anacrolix/torrent   torrent.go:364 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Opening connection with peer 10.65.170.47:7191...  [ github.com/anacrolix/torrent   torrent.go:354 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Inside openNewConns() function  [ github.com/anacrolix/torrent   torrent.go:1576 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: t.peers.Len(): 1  [ github.com/anacrolix/torrent   torrent.go:1578 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: t.peers.PopMax(): 10.65.170.47:7191  [ github.com/anacrolix/torrent   torrent.go:1597 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: initiating connection with peer 10.65.170.47:7191 with opts: {{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.170.47:7191 M false 0 true} 0xc000208a88 false false false {false true}}  [ github.com/anacrolix/torrent   torrent.go:1606 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Inside initateConn() with opts: {{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.170.47:7191 M false 0 true} 0xc000208a88 false false false {false true}}  [ github.com/anacrolix/torrent   torrent.go:2730 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Not ignoring Limits since !ignoreLimits true  [ github.com/anacrolix/torrent   torrent.go:2743 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: calling go t.cl.outgoingConnection({{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.170.47:7191 M false 0 true} 0xc000208a88 false false false {false true}}, &{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.170.47:7191 M false 0 true})  [ github.com/anacrolix/torrent   torrent.go:2756 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Connection initiated with peer 10.65.170.47:7191 with opts: {{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.170.47:7191 M false 0 true} 0xc000208a88 false false false {false true}}  [ github.com/anacrolix/torrent   torrent.go:1609 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Final value of added: true for peer 0xa80660  [ github.com/anacrolix/torrent   torrent.go:364 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Opening connection with peer 10.65.140.160:7191...  [ github.com/anacrolix/torrent   torrent.go:354 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Inside openNewConns() function  [ github.com/anacrolix/torrent   torrent.go:1576 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: t.peers.Len(): 1  [ github.com/anacrolix/torrent   torrent.go:1578 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: t.peers.PopMax(): 10.65.140.160:7191  [ github.com/anacrolix/torrent   torrent.go:1597 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: initiating connection with peer 10.65.140.160:7191 with opts: {{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.140.160:7191 M false 0 true} 0xc000208a88 false false false {false true}}  [ github.com/anacrolix/torrent   torrent.go:1606 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Inside initateConn() with opts: {{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.140.160:7191 M false 0 true} 0xc000208a88 false false false {false true}}  [ github.com/anacrolix/torrent   torrent.go:2730 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Inside outgoingConnection() %!v(MISSING)  [ github.com/anacrolix/torrent   client.go:887 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Inside outgoingConnection() %!v(MISSING)  [ github.com/anacrolix/torrent   client.go:887 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Not ignoring Limits since !ignoreLimits true  [ github.com/anacrolix/torrent   torrent.go:2743 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: calling go t.cl.outgoingConnection({{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.140.160:7191 M false 0 true} 0xc000208a88 false false false {false true}}, &{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.140.160:7191 M false 0 true})  [ github.com/anacrolix/torrent   torrent.go:2756 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Connection initiated with peer 10.65.140.160:7191 with opts: {{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.140.160:7191 M false 0 true} 0xc000208a88 false false false {false true}}  [ github.com/anacrolix/torrent   torrent.go:1609 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Final value of added: true for peer 0xa80660  [ github.com/anacrolix/torrent   torrent.go:364 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Opening connection with peer 10.65.138.33:7191...  [ github.com/anacrolix/torrent   torrent.go:354 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Inside openNewConns() function  [ github.com/anacrolix/torrent   torrent.go:1576 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: t.peers.Len(): 1  [ github.com/anacrolix/torrent   torrent.go:1578 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: t.peers.PopMax(): 10.65.138.33:7191  [ github.com/anacrolix/torrent   torrent.go:1597 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: initiating connection with peer 10.65.138.33:7191 with opts: {{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.138.33:7191 M false 0 true} 0xc000208a88 false false false {false true}}  [ github.com/anacrolix/torrent   torrent.go:1606 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Inside initateConn() with opts: {{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.138.33:7191 M false 0 true} 0xc000208a88 false false false {false true}}  [ github.com/anacrolix/torrent   torrent.go:2730 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Not ignoring Limits since !ignoreLimits true  [ github.com/anacrolix/torrent   torrent.go:2743 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: calling go t.cl.outgoingConnection({{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.138.33:7191 M false 0 true} 0xc000208a88 false false false {false true}}, &{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.138.33:7191 M false 0 true})  [ github.com/anacrolix/torrent   torrent.go:2756 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Connection initiated with peer 10.65.138.33:7191 with opts: {{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.138.33:7191 M false 0 true} 0xc000208a88 false false false {false true}}  [ github.com/anacrolix/torrent   torrent.go:1609 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Final value of added: true for peer 0xa80660  [ github.com/anacrolix/torrent   torrent.go:364 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Opening connection with peer 10.65.131.12:7191...  [ github.com/anacrolix/torrent   torrent.go:354 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: Inside openNewConns() function  [ github.com/anacrolix/torrent   torrent.go:1576 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: t.peers.Len(): 1  [ github.com/anacrolix/torrent   torrent.go:1578 ] 

2024-07-09T00:22:40.175-0700	debug	log/logger.go:157	#Floodix: t.peers.PopMax(): 10.65.131.12:7191  [ github.com/anacrolix/torrent   torrent.go:1597 ] 

2024-07-09T00:22:40.176-0700	debug	log/logger.go:157	#Floodix: initiating connection with peer 10.65.131.12:7191 with opts: {{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.131.12:7191 M false 0 true} 0xc000208a88 false false false {false true}}  [ github.com/anacrolix/torrent   torrent.go:1606 ] 

2024-07-09T00:22:40.176-0700	debug	log/logger.go:157	#Floodix: Inside initateConn() with opts: {{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.131.12:7191 M false 0 true} 0xc000208a88 false false false {false true}}  [ github.com/anacrolix/torrent   torrent.go:2730 ] 

2024-07-09T00:22:40.176-0700	debug	log/logger.go:157	#Floodix: Not ignoring Limits since !ignoreLimits true  [ github.com/anacrolix/torrent   torrent.go:2743 ] 

2024-07-09T00:22:40.176-0700	debug	log/logger.go:157	#Floodix: calling go t.cl.outgoingConnection({{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.131.12:7191 M false 0 true} 0xc000208a88 false false false {false true}}, &{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.131.12:7191 M false 0 true})  [ github.com/anacrolix/torrent   torrent.go:2756 ] 

2024-07-09T00:22:40.176-0700	debug	log/logger.go:157	#Floodix: Connection initiated with peer 10.65.131.12:7191 with opts: {{[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] 10.65.131.12:7191 M false 0 true} 0xc000208a88 false false false {false true}}  [ github.com/anacrolix/torrent   torrent.go:1609 ] 

2024-07-09T00:22:40.176-0700	debug	log/logger.go:157	#Floodix: Final value of added: true for peer 0xa80660  [ github.com/anacrolix/torrent   torrent.go:364 ] 

2024-07-09T00:22:40.176-0700	info	floodix/server.go:324	Added 5 new peers
```

but it seems this is not working due to other peers2..5 logging above mentioned log. 

Can you please suggest why the node2 is logging 
```
2024-07-09T00:22:54.305-0700	debug	log/logger.go:157	#Floodix: torrent doesn't want outgoing connections since !t.wantOutgoingConns(): true  [ github.com/anacrolix/torrent   torrent.go:1581 ] 
```

abhishek-das-gupta commented:
Hi @anacrolix , Is there a client config which lets me change the download request size when downloading from web seed?  Like suppose I want to request 1 MiB at a time from webseed?

anacrolix commented:
The webseeding code mirrors the regular code in the size of chunks it attempts. Look for "chunksize" in the client or torrent config. I know a few users set it very high (1MiB or so). There's no way to set it just for webseeding as webseeds and regular conns share the same request algorithm.

abhishek-das-gupta commented:
Thanks, @anacrolix, for your test case! Using your testcase, once I set the optimal piece length, the torrent completed quickly. However, I have one question: suppose there is no seeder to begin with (another use case). My fallback is to contact the webseed to download the file.

All the peers within the swarm contact the single webseed, but I was hoping they would also communicate among each other to distribute the pieces. However, it seems like this is not happening, as I don't see any activity under `requested pieces:`. Please see the following status of peer #4. You can see that it is not requesting any pieces from its peers, resulting in a sequential HTTP download.
```
<some-parcel>.parcel
16.575733% of 4048621206 bytes (4.0 GB)
Infohash: 075b6c84d4bb6e83046f6b14c77b075de307e318
Metadata length: 38741
Piece length: 2097152 (128 chunks)
Num Pieces: 1931 (320 completed)
Piece States: 2C 1. 2C 1. 8C 1. 1C 1. 4C 1. 4C 1. 5C 1. 1C 1. 18C 1. 3C 1. 5C 1. 5C 1. 6C 1. 7C 1. 1C 1. 6C 1. 31C 1. 18C 1. 10C 1. 8C 1. 3C 1. 4C 1. 9C 1. 23C 1. 51C 1. 15C 1. 12C 1. 22C 1. 15C 1. 1C 1. 20C 1581.
Piece availability frequency: 0: 1611, 4: 320
Reader Pieces:
Enabled trackers:
    URL  Extra
DHT Announces: 0
(torrent.TorrentStats) {
 ConnStats: (torrent.ConnStats) {
  BytesWritten: (torrent.Count) 288237498,
  BytesWrittenData: (torrent.Count) 287277056,
  BytesRead: (torrent.Count) 672103818,
  BytesReadData: (torrent.Count) 671219712,
  BytesReadUsefulData: (torrent.Count) 671088640,
  BytesReadUsefulIntendedData: (torrent.Count) 671088640,
  ChunksWritten: (torrent.Count) 17534,
  ChunksRead: (torrent.Count) 40968,
  ChunksReadUseful: (torrent.Count) 40960,
  ChunksReadWasted: (torrent.Count) 8,
  MetadataChunksRead: (torrent.Count) 0,
  PiecesDirtiedGood: (torrent.Count) 320,
  PiecesDirtiedBad: (torrent.Count) 0
 },
 TotalPeers: (int) 4,
 PendingPeers: (int) 0,
 ActivePeers: (int) 4,
 ConnectedSeeders: (int) 0,
 HalfOpenPeers: (int) 0,
 PiecesComplete: (int) 320
}
webseeds:
4 peer conns:
- 10.65.55.225:7191-10.65.54.157:52422
  peer id: "-GT0003-[\t\x1b\xb6\x00\xeaa\xaf\x97h\x9a\xec"
  extensions: 0000000000100005 (ltep, fast, dht)
  ltep extensions: map[ut_holepunch:2 ut_metadata:1 ut_pex:3]
  pex: 4 conns, 0 unsent events
  bep40-prio: f3cfa364
  last msg: 0.11s ago, connected: 84.55s ago, last helpful: 0.11s ago, itime: 1m24.544555365s, etime: 4.811743861s
  320/1931 completed, 0 pieces touched, good chunks: 40561/40561:0 reqq: 0+0/(1/1024):0/1024, flags: i:I,e,v1:, dr: 134873.3 KiB/s
  requested pieces:
- 0.0.0.0:7191-10.65.51.253:7191
  peer id: "-GT0003-_-6\x8b\xac^\x95\x83\x06x\xdbA"
  extensions: 0000000000100005 (ltep, fast, dht)
  ltep extensions: map[ut_holepunch:2 ut_metadata:1 ut_pex:3]
  pex: 4 conns, 0 unsent events
  bep40-prio: ef9c4773
  last msg: 0.10s ago, connected: 84.60s ago, last helpful: 20.98s ago, itime: 578.115117ms, etime: 486.698749ms
  320/1931 completed, 0 pieces touched, good chunks: 23/31:272 reqq: 0+0/(1/1024):0/1024, flags: :I,U,e,v1:i, dr: 756.1 KiB/s
  requested pieces:
- 10.65.55.225:7191-10.65.49.106:54548
  peer id: "-GT0003-\x93D6\x8cN\x99$%E\x04\xaf\x1d"
  extensions: 0000000000100005 (ltep, fast, dht)
  ltep extensions: map[ut_holepunch:2 ut_metadata:1 ut_pex:3]
  pex: 4 conns, 0 unsent events
  bep40-prio: ca339400
  last msg: 0.10s ago, connected: 84.60s ago, last helpful: 13.01s ago, itime: 1m24.434434388s, etime: 65.271288ms
  320/1931 completed, 0 pieces touched, good chunks: 254/254:1335 reqq: 0+0/(1/1024):0/1024, flags: i:I,e,v1:i, dr: 62263.2 KiB/s
  requested pieces:
- 10.65.55.225:7191-10.65.50.36:46832
  peer id: "-GT0003-\x8a\xfd\xf8\xaf\x95\xbce\xf1h=\x82\xd2"
  extensions: 0000000000100005 (ltep, fast, dht)
  ltep extensions: map[ut_holepunch:2 ut_metadata:1 ut_pex:3]
  pex: 4 conns, 0 unsent events
  bep40-prio: 5f7086d4
  last msg: 0.07s ago, connected: 84.59s ago, last helpful: 0.09s ago, itime: 1m9.563634782s, etime: 9.983955ms
  320/1931 completed, 0 pieces touched, good chunks: 122/122:15927 reqq: 0+0/(1/1024):0/1024, flags: i:I,e,v1:i, dr: 195513.7 KiB/s
  requested pieces:
```

When I switch to  Libtorrent, I see that peers simultaneously download from the webseed and distribute pieces among themselves.

Is there a way to achieve this in your torrent library?

anacrolix commented:
It looks like your peers are downloading in lockstep. If you look at their "x/y completed", they're all at the same value. They're sharing faster than it takes to request more pieces (I'm guessing they're on the same local network or host). The requested pieces field shows active requests. At the time the status was generated there are no active requests. If you look at "good chunks" that shows how many requests were satisfied from the peer since they last connected (and how many they uploaded to that same peer).

anacrolix commented:
Thanks for working through this, and the contribution!
